modular_zubbers/code/modules/storyteller/antagonist_event.dm

```diff
@@ -0,0 +1,170 @@
+	/// The minds we've setup in setup() and need to finalize in start()
+	var/list/setup_minds = list()
+	/// Whether we prompt the players before picking them.
+	var/prompted_picking = FALSE //TODO: Implement this
+
+/datum/round_event/antagonist/solo/setup()
+	var/datum/round_event_control/antagonist/solo/cast_control = control
+	antag_count = cast_control.get_antag_amount()
+	antag_flag = cast_control.antag_flag
+	antag_datum = cast_control.antag_datum
+	restricted_roles = cast_control.restricted_roles
+	prompted_picking = cast_control.prompted_picking
+
+	var/list/candidates = cast_control.get_candidates()
+	for(var/i in 1 to antag_count)
+		if(!candidates.len)
+			break
+		var/mob/candidate = pick_n_take(candidates)
+		setup_minds += candidate.mind
+		candidate.mind.special_role = antag_flag
+		candidate.mind.restricted_roles = restricted_roles
+
+/datum/round_event/antagonist/solo/start()
+	for(var/datum/mind/antag_mind as anything in setup_minds)
+		add_datum_to_mind(antag_mind)
+
+/datum/round_event/antagonist/solo/proc/add_datum_to_mind(datum/mind/antag_mind)
+	antag_mind.add_antag_datum(antag_datum)
+
+/datum/round_event_control/antagonist/solo/traitor
+	antag_flag = ROLE_TRAITOR
+	antag_datum = /datum/antagonist/traitor
+	protected_roles = list("Prisoner","Security Officer", "Warden", "Detective", "Head of Security", "Captain")
+	restricted_roles = list("AI", "Cyborg")
+
+/datum/round_event_control/antagonist/solo/traitor/roundstart
+	name = "Traitors"
+	roundstart = TRUE
+
+/datum/round_event_control/antagonist/solo/traitor/midround
+	name = "Sleeper Agents (Traitors)"
+	prompted_picking = TRUE
+
+/datum/round_event_control/antagonist/solo/changeling
+	antag_flag = ROLE_CHANGELING
+	antag_datum = /datum/antagonist/changeling
+	protected_roles = list("Prisoner","Security Officer", "Warden", "Detective", "Head of Security", "Captain")
+	restricted_roles = list("AI", "Cyborg")
+	weight = 4
+	min_players = 20
+
+/datum/round_event_control/antagonist/solo/changeling/roundstart
+	name = "Changelings"
+	roundstart = TRUE
+
+/datum/round_event_control/antagonist/solo/changeling/midround
+	name = "Genome Awakening (Changelings)"
+	prompted_picking = TRUE
+
+/datum/round_event_control/antagonist/solo/heretic
+	antag_flag = ROLE_HERETIC
+	antag_datum = /datum/antagonist/heretic
+	protected_roles = list("Prisoner","Security Officer", "Warden", "Detective", "Head of Security", "Captain")
+	restricted_roles = list("AI", "Cyborg")
+	weight = 4
+	min_players = 20
+
+/datum/round_event_control/antagonist/solo/heretic/roundstart
+	name = "Heretics"
+	roundstart = TRUE
+
+/datum/round_event_control/antagonist/solo/heretic/midround
+	name = "Midround Heretics"
+	prompted_picking = TRUE
+
+/datum/round_event_control/antagonist/solo/wizard
+	name = "Wizard"
+	typepath = /datum/round_event/antagonist/solo/wizard
+	antag_flag = ROLE_WIZARD
+	antag_datum = /datum/antagonist/wizard
+	restricted_roles = list("Head of Security", "Captain")
```

Comment:
Is this a good idea considering that wizard is an offstation role or is this needed to prevent empty slots?

Reply:
Wizard is not a station-spawned role, aka what Burger said - wouldnt this just do nothing, given that antagonist roles roll first before job picks?

Reply:
I'm confused about this code myself, if I sng up as, say, secoff, will I spawn in with my gear and then get teleported to the wizards den?

Reply:
You can just run freejobslot on the job chosen afterwards, but heads are bad IMO because it announces most heads and i think it would make offstations instantly metagameable like abductors or non war nukies. Also im in the boat of running these after roles are setup so we can have proper control over roundstart events based on made occupations, IE no nukies without x number of security officers and atleast one head